### PR TITLE
✨ [file based cdk] S3 legacy config adapter

### LINKED ISSUE #28131: File CDK: S3 config adapter (top-level config options)
The existing S3 connector offers some config options that we do not want to provide for new connections. However, we still want to support them rather than making changes that will break users' syncs.

At the top-level, the options provided by the S3 connector include:
- `dataset`: should be mapped to `name`.
- `path_pattern`: should be mapped to `globs`. I don't think additional changes are required but we should double-check.
- `path_prefix`: we should attempt to create the appropriate globs by combining this with `path_pattern`.
- `user_schema`: should be mapped to `input_schema`.
- `start_date`: this will be wired through the file-based CDK in https://github.com/airbytehq/airbyte/issues/28136

This will also include handling [config options used by the existing S3 connector](https://github.com/airbytehq/airbyte/blob/af4b0d7244ef1002c13b8c267326be3e9973f6f7/airbyte-integrations/connectors/source-s3/source_s3/source.py#L15) including:
- `bucket`
- `aws_access_key_id`
- `aws_secret_access_key`
- `endpoint`

These options are currently nested under a `provider` key (along with `path_prefix` and `start_date`), but for the new S3 connector we'll have them be at the top level.

This ticket will involve 1) adding a version to the new file-based CDK config object so that we can differentiate it from the existing S3 config and 2) creating a new S3-specific FileBasedConfig object, to be used by the new S3 connector. The new config object will include logic to adapt the old config to the new config format. In addition to top-level config options, this pattern will be used by the config options specific to each file type (described in separate tickets).

### PR DESCRIPTION
**Pull request by brianjlai**
Closes #28131

## What
Adds a transformer that allows the new S3 connector built with the file-based CDK to accept configs in the legacy format to be processed. It does not persist the change in the DB, just how the connector handles an operation

## How

The CDK allows for a source to override the `read_config()` method which is responsible for reading in the config JSON file. By overriding this, we can still get the JSON from the file, but we also then parse it into the legacy format instance and then construct a new dictionary representing the new format. By doing it this way, the downstream flow can assume it is always using the new config

## Recommended reading order
1. `legacy_config_transformer.py`
2. `s3_source.py`

### CODE REVIEW

**Comment by octavia-squidington-iii** [comment]
##  source-s3 test report (commit [`ac88970854`](https://github.com/airbytehq/airbyte/pull/29145/commits/ac88970854bad2827350f354d4306392e198ac50)) - ❌

⏲️  Total pipeline duration: 22mn01s 

| Step                                                             | Result   |
|:-----------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-s3/metadata.yaml | ✅        |
| Connector version semver check                                   | ✅        |
| Connector version increment check                                | ❌        |
| QA checks                                                        | ✅        |
| Code format checks                                               | ❌        |
| Connector package install                                        | ✅        |
| Build source-s3 docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                       | ✅        |
| Integration tests                                                | ✅        |
| Acceptance tests                                                 | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/brian_s3_file_based_adapter/1691429057/ac88970854bad2827350f354d4306392e198ac50/source-s3/3.1.2/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:ac88970854bad2827350f354d4306392e198ac50)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```

File: airbyte-integrations/connectors/source-s3/source_s3/v4/legacy_config_transformer.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+from source_s3.source import SourceS3Spec
+from typing import Mapping, Any, List
+
+
+class LegacyConfigTransformer:
```
Comment (girarda) [inline airbyte-integrations/connectors/source-s3/source_s3/v4/legacy_config_transformer.py:9]: nit: can you add a comment in the spec that this file needs to be updated if the spec changes? I'm not sure how long we'll need this transformer and there are PRs in the pipeline (eg https://github.com/airbytehq/airbyte/pull/17334)

**Review by girarda — APPROVED** [review]

File: airbyte-integrations/connectors/source-s3/source_s3/v4/legacy_config_transformer.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            transformed_config["aws_secret_access_key"] = legacy_config.provider.aws_secret_access_key
+        if legacy_config.provider.endpoint:
+            transformed_config["endpoint"] = legacy_config.provider.endpoint
+        if legacy_config.user_schema and legacy_config.user_schema != "{}":
+            transformed_config["streams"][0]["input_schema"] = legacy_config.user_schema
+
+        return transformed_config
+
+    @ classmethod
+    def create_globs(cls, path_pattern: str, path_prefix: str) -> List[str]:
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-s3/source_s3/v4/legacy_config_transformer.py:44]: nit: should we make this private?

**Review by maxi297 — APPROVED** [review]

**Review by clnoll — APPROVED** [review]
Nice @brianjlai!

File: airbyte-integrations/connectors/source-s3/source_s3/v4/legacy_config_transformer.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    @classmethod
+    def convert(cls, legacy_config: SourceS3Spec) -> Mapping[str, Any]:
+        transformed_config = {
+            "bucket": legacy_config.provider.bucket,
+            "streams": [
+                {
+                    "name": legacy_config.dataset,
+                    "file_type": legacy_config.format.filetype,
+                    "globs": cls.create_globs(legacy_config.path_pattern, legacy_config.provider.path_prefix),
+                    "validation_policy": "emit_record",
```
Comment (girarda) [inline airbyte-integrations/connectors/source-s3/source_s3/v4/legacy_config_transformer.py:24]:
```suggestion
                    "validation_policy": "Emit Record",
```

Comment (brianjlai) [reply airbyte-integrations/connectors/source-s3/source_s3/v4/legacy_config_transformer.py:24]: yep fixed that after I merged in the latest master

File: airbyte-integrations/connectors/source-s3/source_s3/v4/legacy_config_transformer.py (hunk 1 above)
Comment (brianjlai) [reply airbyte-integrations/connectors/source-s3/source_s3/v4/legacy_config_transformer.py:14]: done!

**Comment by octavia-squidington-iii** [comment]
##  source-s3 test report (commit [`a6159c1cbe`](https://github.com/airbytehq/airbyte/pull/29145/commits/a6159c1cbed13cc1973e8efa55d93b88a3fe57ce)) - ❌

⏲️  Total pipeline duration: 21mn56s 

| Step                                                             | Result   |
|:-----------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-s3/metadata.yaml | ✅        |
| Connector version semver check                                   | ✅        |
| Connector version increment check                                | ❌        |
| QA checks                                                        | ✅        |
| Code format checks                                               | ❌        |
| Connector package install                                        | ✅        |
| Build source-s3 docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                       | ✅        |
| Integration tests                                                | ✅        |
| Acceptance tests                                                 | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/brian_s3_file_based_adapter/1691614653/a6159c1cbed13cc1973e8efa55d93b88a3fe57ce/source-s3/3.1.2/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:a6159c1cbed13cc1973e8efa55d93b88a3fe57ce)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```

**Comment by octavia-squidington-iii** [comment]
##  source-s3 test report (commit [`5ef32c6e2d`](https://github.com/airbytehq/airbyte/pull/29145/commits/5ef32c6e2dd833ddcbb1eb1e2e2e04df0d6382cf)) - ❌

⏲️  Total pipeline duration: 22mn19s 

| Step                                                             | Result   |
|:-----------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-s3/metadata.yaml | ✅        |
| Connector version semver check                                   | ✅        |
| Connector version increment check                                | ❌        |
| QA checks                                                        | ✅        |
| Code format checks                                               | ❌        |
| Connector package install                                        | ✅        |
| Build source-s3 docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                       | ✅        |
| Integration tests                                                | ✅        |
| Acceptance tests                                                 | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/brian_s3_file_based_adapter/1691616065/5ef32c6e2dd833ddcbb1eb1e2e2e04df0d6382cf/source-s3/3.1.2/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:5ef32c6e2dd833ddcbb1eb1e2e2e04df0d6382cf)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```

**Comment by octavia-squidington-iii** [comment]
##  source-s3 test report (commit [`391c1ee09d`](https://github.com/airbytehq/airbyte/pull/29145/commits/391c1ee09d2978d47b5ad42fde8d3d1d6348814a)) - ❌

⏲️  Total pipeline duration: 22mn00s 

| Step                                                             | Result   |
|:-----------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-s3/metadata.yaml | ✅        |
| Connector version semver check                                   | ✅        |
| Connector version increment check                                | ❌        |
| QA checks                                                        | ✅        |
| Code format checks                                               | ✅        |
| Connector package install                                        | ✅        |
| Build source-s3 docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                       | ✅        |
| Integration tests                                                | ✅        |
| Acceptance tests                                                 | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/brian_s3_file_based_adapter/1691619415/391c1ee09d2978d47b5ad42fde8d3d1d6348814a/source-s3/3.1.2/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:391c1ee09d2978d47b5ad42fde8d3d1d6348814a)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```

**Pull request merged**